edge/cmd/edgemark/hollow_edgecore.go

```diff
@@ -0,0 +1,157 @@
+func newHollowEdgeNodeCommand() *cobra.Command {
+	s := &hollowEdgeNodeConfig{
+		NodeLabels: make(map[string]string),
+	}
+
+	cmd := &cobra.Command{
+		Use:  "edgemark",
+		Long: "edgemark",
+		Run: func(cmd *cobra.Command, args []string) {
+			verflag.PrintAndExitIfRequested()
+			run(s)
+		},
+		Args: func(cmd *cobra.Command, args []string) error {
+			for _, arg := range args {
+				if len(arg) > 0 {
+					return fmt.Errorf("%q does not take any arguments, got %q", cmd.CommandPath(), args)
+				}
+			}
+			return nil
+		},
+	}
+
+	fs := cmd.Flags()
+	fs.AddGoFlagSet(flag.CommandLine) // for flags like --docker-only
+	s.addFlags(fs)
+
+	return cmd
+}
+
+func run(config *hollowEdgeNodeConfig) {
+	c := EdgeCoreConfig(config)
+
+	// use fake runtime service
+	edged.DefaultGetRuntimeService = GetFakeRuntimeAndImageServices
+
+	edged.Register(c.Modules.Edged)
+	edgehub.Register(c.Modules.EdgeHub, c.Modules.Edged.HostnameOverride)
+	metamanager.Register(c.Modules.MetaManager)
+
+	dbm.InitDBConfig(c.DataBase.DriverName, c.DataBase.AliasName, c.DataBase.DataSource)
+
+	// start all modules
+	core.Run()
+}
+
+func (c *hollowEdgeNodeConfig) addFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&c.Token, "token", "", "Token indicates the priority of joining the cluster for the edge.")
+	fs.StringVar(&c.NodeName, "name", "fake-node", "Name of this Hollow Node.")
+	fs.StringVar(&c.WebsocketServer, "websocket-server", "", "Server indicates websocket server address.")
+	fs.StringVar(&c.HTTPServer, "http-server", "", "HTTPServer indicates the server for edge to apply for the certificate.")
+	bindableNodeLabels := cliflag.ConfigurationMap(c.NodeLabels)
+	fs.Var(&bindableNodeLabels, "node-labels", "Additional node labels")
+}
+
+func EdgeCoreConfig(config *hollowEdgeNodeConfig) *v1alpha1.EdgeCoreConfig {
+	edgeCoreConfig := v1alpha1.NewDefaultEdgeCoreConfig()
+
+	// overWrite config
+	edgeCoreConfig.DataBase.DataSource = "/edgecore.db"
+	edgeCoreConfig.Modules.EdgeHub.Token = config.Token
+	edgeCoreConfig.Modules.EdgeHub.HTTPServer = config.HTTPServer
+	edgeCoreConfig.Modules.EdgeHub.WebSocket.Server = config.WebsocketServer
+
+	// use fake runtime for test
+	edgeCoreConfig.Modules.Edged.RuntimeType = "fake"
+	edgeCoreConfig.Modules.Edged.RemoteRuntimeEndpoint = "/run/fake/fake.sock"
+	edgeCoreConfig.Modules.Edged.RemoteImageEndpoint = "/run/fake/fake.sock"
+	edgeCoreConfig.Modules.Edged.EnableMetrics = false
+
+	edgeCoreConfig.Modules.Edged.HostnameOverride = config.NodeName
+	edgeCoreConfig.Modules.Edged.Labels = config.NodeLabels
+
+	return edgeCoreConfig
+}
+
+func GetFakeRuntimeAndImageServices(
+	remoteRuntimeEndpoint,
+	remoteImageEndpoint string,
+	runtimeRequestTimeout metav1.Duration) (internalapi.RuntimeService, internalapi.ImageManagerService, error) {
+	endpoint, err := fakeremote.GenerateEndpoint()
+	if err != nil {
+		klog.Errorf("Failed to generate fake endpoint %v.", err)
+		return nil, nil, err
+	}
```

Comment:
```suggestion
	if err != nil {
		return nil, nil, fmt.Errorf("Failed to generate fake endpoint: %v.", err)
	}
```
```golang
func Register(e *v1alpha1.Edged) {
	edgedconfig.InitConfigure(e)
	edged, err := newEdged(e.Enable)
	if err != nil {
		klog.Errorf("init new edged error, %v", err)
		os.Exit(1)
	}
	core.Register(edged)
}
```
After `edged` initialization fails, an error will be logged. Wouldn't it be better to use `fmt.Errorf()` here?

Reply:
done